C#/Botkit/Microsoft.BotKit/BotWorker.cs

```diff
@@ -118,25 +149,79 @@ public async Task<DialogTurnResult> CancelAllDialogs()
         /// </summary>
         /// <param name="id">ID of dialog</param>
         /// <param name="options">Object containing options to be passed into the dialog</param>
-        public async Task ReplaceDialog(string id, object options)
+        public async Task ReplaceDialog(string id, Dictionary<string, object> options)
         {
+            if (config.DialogContext != null)
+            {
+                var opt = new Dictionary<string, object>();
+
+                // TO-DO: Review this implementation of 'Options' as Dictionary
+                foreach (KeyValuePair<string, object> entry in options)
+                {
+                    opt.Add(entry.Key, entry.Value);
+                }
 
+                opt.Add("user", config.TurnContext.Activity.From.Id);
+                opt.Add("channel", config.TurnContext.Activity.Conversation.Id);
+
+                await config.DialogContext.ReplaceDialogAsync(id + ":botkit-wrapper", opt);
+
+                // make sure we save the state change caused by the dialog.
+                // this may also get saved again at end of turn
+                await Controller.SaveState(this);
+            }
         }
 
         /// <summary>
         /// Alter the context in which a bot instance will send messages.
         /// Use this method to create or adjust a bot instance so that it can send messages to a predefined user/channel combination.
         /// </summary>
-        /// <param name="reference">A ConversationReference, most likely captured from an incoming message and stored for use in proactive messaging scenarios.</param>
-        public async Task<BotWorker> ChangeContext(ConversationReference reference)
+        /// <param name="conversationReference">A ConversationReference, most likely captured from an incoming message and stored for use in proactive messaging scenarios.</param>
+        public async Task<BotWorker> ChangeContext(ConversationReference conversationReference)
         {
-            return new BotWorker(new Botkit(new BotkitConfiguration()), new object());
-        }
+            // change context of outbound activities to use this new address
+            config.ConversationReference = conversationReference;
+
+            // Create an activity using this reference
+            var activity = new Activity();
+            activity.ApplyConversationReference(conversationReference, true);
```

Comment:
Here you should do this:
```
activity = activity.ApplyConversationReference(conversationReferemce, true)
```
As if you left these lines as they actually are, the activity won't be updated.

BTW, this is the current implementation of _ApplyConversation_ method in _Activity_ class ([link](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Schema/ActivityEx.cs#L473
)):

![image](https://user-images.githubusercontent.com/43762887/58653395-fcb29c00-82eb-11e9-9aca-0c24f83b5a1f.png)